test/mpi/errors/comm/unmatched.c

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpitest.h"
+
+int main(void)
+{
+    int mpi_errno = MPI_SUCCESS;
+    int errs = 0;
+
+    MTest_Init(NULL, NULL);
+
+    MPI_Comm comm;
+    MPI_Comm_dup(MPI_COMM_WORLD, &comm);
+    MPI_Comm_set_errhandler(comm, MPI_ERRORS_RETURN);
+    /* The implementation may raise an error on the default handler
+     * since it will happen during MPI_Comm_free. The default,
+     * as of MPI-4.0, is the error handler for MPI_COMM_SELF.
+     * Prior to MPI-4.0, it was MPI_COMM_WORLD. */
+    MPI_Comm_set_errhandler(MPI_COMM_SELF, MPI_ERRORS_RETURN);
+    MPI_Comm_set_errhandler(MPI_COMM_WORLD, MPI_ERRORS_RETURN);
+
+    int rank;
+    MPI_Comm_rank(comm, &rank);
+
+    if (rank == 0) {
+        /* if an implementation does not use eager
+         * protocol, this send may hang */
+        MPI_Send(NULL, 0, MPI_INT, 1, 0, comm);
+    }
+
+    MPI_Barrier(comm);
+
+    mpi_errno = MPI_Comm_free(&comm);
+    /* unmatched message should be detected at rank 1 */
+    if (rank == 1 && mpi_errno == MPI_SUCCESS) {
+        errs++;
+    }
```

Comment:
Thinking about this test and I think the check in `MPI_COMM_FREE` has a flaw. What if I have inactive persistent recv requests that could match the messages? Such a request could be started after the free 🤔.

Reply:
Yeah, the check should be at comm release time (i.e. when ref_count reaches 0), not MPI_Comm_free. I guess what's missing is checking remaining non-builtin `MPIR_Comm` at MPI_Finalize or MPI_Session_finalize, and garbage-collect the comms.

Reply:
Digging back into this a little bit, the check is in fact after the ref_count has reached zero. So I think the test as it is should work since there are no additional references to the comm at the time of the free. Still, using disconnect might be a more robust test of the functionality. Will submit a patch for consideration.